daemon/main.go

```diff
@@ -810,6 +811,53 @@ func runDaemon() {
 	log.Info("Launching node monitor daemon")
 	go d.nodeMonitor.Run(path.Join(defaults.RuntimePath, defaults.EventsPipe), bpf.GetMapRoot())
 
+	eventsCh, err := workloads.EnableEventListener()
+	if err != nil {
+		log.WithError(err).Fatal("Error while enabling docker event watcher")
+	} else {
+		d.workloadsEventsCh = eventsCh
+	}
+
+	if err := d.EnableK8sWatcher(5 * time.Minute); err != nil {
+		log.WithError(err).Fatal("Unable to establish connection to Kubernetes apiserver")
+	}
+
+	if option.Config.RestoreState {
+		// Block on regenerating endpoints until we receive all policies from K8s
+		// if K8s is enabled.
+		if k8s.IsEnabled() {
+			log.Info("waiting to regenerate restored endpoints until all policies have been synced from Kubernetes")
+			d.k8sResourceSyncWaitGroup.Wait()
```

Comment:
@tgraf @aanm the side effect of this implementation is that it blocks the serving of the API until all initial NP and CNP are read from K8s; I can work on making this an async process to maintain similar startup time as before? Maybe we can have a discussion on what is required for us to serve the API for Cilium (i.e., do all endpoints have to be restored, k8s watcher started, etc.). We can document this somewhere and add it to the code accordingly to ensure that we maintain said behavior over time.

Reply:
I think blocking the API is fine. This should be a quite fast operation. I think we can discuss whether we want to serve the status API earlier than the endpoint API but we definitely don't want to allow creating endpoints before the policy is known.